### PR TITLE
Prevents changing the parent of an entity in a prefab

### LINKED ISSUE #7670: Editor crashes after undoing the parent change on an entity inside a Prefab
**Describe the bug**
When an entity's parent change is undone on an entity inside a **Prefab** the Editor crashes.
The issue only occurs when the new parent entity is outside the **Prefab**, changing the parent of the entity in a **Prefab** to another entity present in that **Prefab** and undoing the action will **not** result in the Editor crashing.

Please note that after the entity's parent is changed, multiple warnings and error are printed to the console (please refer to the Editor.log file).

It has been verified that this issue occurs on both Linux and Windows systems.

**Steps to reproduce**
1. Create a **Prefab**.
2. Enter the Focus mode of the created **Prefab**.
3. Select the entity inside the **Prefab**.
4. In the Entity Inspector, change the **Parent Entity** to an entity outside of the prefab.
5. Undo the action (Ctrl + Z).

**Expected behavior**
The action of changing the parent of the entity inside a **Prefab** can be undone, the Editor remains stable.

**Actual behavior**
Undoing the parent change on an entity inside a **Prefab** crashes the Editor.
 

**Video**

https://user-images.githubusercontent.com/86951967/154258377-dc733e8d-ab52-48b0-876c-a20ee02afe08.mp4

**Found in Branch**
Development (16d6cbc)

**Desktop:**
 - Device: PC
 - OS: Linux
 - Version: Ubuntu 20.04.3 LTS
 - CPU: Intel Core i7-8700K
 - GPU: NVIDIA GeForce RTX 2080 Ti
 - Memory: 16GB

**Additional context**
[Editor.log](https://github.com/o3de/o3de/files/8079562/Editor.log)

### PR DESCRIPTION
**Pull request by jjjoness**
Fixes #7670

Signed-off-by: John Jones-Steele <82226755+jjjoness@users.noreply.github.com>

At present it is illegal to move a child entity of a prefab into another container. Until that is fixed, this code prevents it happening.

### CODE REVIEW

File: Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp [hunk 1]
```diff
@@ -194,6 +194,7 @@ namespace AzToolsFramework
             , m_suppressTransformChangedEvent(false)
             , m_interpolatePosition(AZ::InterpolationMode::NoInterpolation)
             , m_interpolateRotation(AZ::InterpolationMode::NoInterpolation)
+            , m_focusModeInterface(AZ::Interface<AzToolsFramework::FocusModeInterface>::Get())
```
Comment (AMZN-daimini) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp:197]: You should check if this interface was actually initialized correctly (if there's no handler, Get will return nullptr)

File: Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp [hunk 2]
```diff
@@ -954,6 +955,13 @@ namespace AzToolsFramework
                 return AZ::Failure(AZStd::string("You cannot set an entity to be a child of a closed container!"));
             }
 
+            // Don't allow entities to be parented outside their container.
+            AzToolsFramework::EntityIdList entities = m_focusModeInterface->GetFocusedEntities(GetEntityContextId());
+            if (std::find(entities.begin(), entities.end(), actualValue) == entities.end())
+            {
+                return AZ::Failure(AZStd::string("You cannot set an entity to be a child of a different container!"));
+            }
```
Comment (AMZN-daimini) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp:963]: nit: This works and shouldn't cause performance concerns (the focused entities list is cached and passed by const ref), but alternatively you can also use IsInFocusSubTree (which just returns true/false)

File: Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp [hunk 3]
```diff
@@ -954,6 +955,13 @@ namespace AzToolsFramework
                 return AZ::Failure(AZStd::string("You cannot set an entity to be a child of a closed container!"));
             }
 
+            // Don't allow entities to be parented outside their container.
+            AzToolsFramework::EntityIdList entities = m_focusModeInterface->GetFocusedEntities(GetEntityContextId());
+            if (std::find(entities.begin(), entities.end(), actualValue) == entities.end())
+            {
+                return AZ::Failure(AZStd::string("You cannot set an entity to be a child of a different container!"));
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp:962]: nit : 'container is a bit too generic'. You can instead say something like "You can only set a parent as one of the entities belonging to the focused prefab"

File: Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp (hunk 1 above)
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp:197]: Added check.

File: Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp (hunk 2 above)
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp]: Much nicer!

File: Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp (hunk 3 above)
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp]: Change made.

File: Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp [hunk 4]
```diff
@@ -954,6 +955,15 @@ namespace AzToolsFramework
                 return AZ::Failure(AZStd::string("You cannot set an entity to be a child of a closed container!"));
             }
 
+            // Don't allow entities to be parented outside their container.
+            if (m_focusModeInterface)
+            {
+                if (!m_focusModeInterface->IsInFocusSubTree(actualValue))
+                {
+                    return AZ::Failure(AZStd::string("You can only set a parent as one of the entities belonging to the focused prefab!"));
```
Comment (srikappa-amzn) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp:963]: nit: remove the '!' at the end to keep the message tone neutral.

**Review by srikappa-amzn — APPROVED** [review]

File: Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp (hunk 4 above)
Comment (jjjoness) [reply Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp]: Removed exclamation marks from all messages in function.

File: Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp [hunk 5]
```diff
@@ -921,37 +922,46 @@ namespace AzToolsFramework
...
                 !containerEntityInterface->IsContainerOpen(actualValue))
             {
-                return AZ::Failure(AZStd::string("You cannot set an entity to be a child of a closed container!"));
+                return AZ::Failure(AZStd::string("You cannot set an entity to be a child of a closed container."));
+            }
+
+            // Don't allow entities to be parented outside their container.
+            if (m_focusModeInterface)
+            {
+                if (!m_focusModeInterface->IsInFocusSubTree(actualValue))
```
Comment (AMZN-daimini) [inline Code/Framework/AzToolsFramework/AzToolsFramework/ToolsComponents/TransformComponent.cpp:961]: nit: can combine into one

```
if (m_focusModeInterface && !m_focusModeInterface->IsInFocusSubTree(actualValue))
```

**Review by AMZN-daimini — APPROVED** [review]

**Pull request merged**